Alternative: restrict Navigation permissions and show UI warning if cannot create

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
In https://github.com/WordPress/gutenberg/pull/37289 we were exploring trying to show a warning if the user cannot "make" new Navigation Menus. However this was proving difficult as whilst even low Role users had permission to `create` they don't have the permission to `publish`. This makes it impossible to show an upfront warning to the user if they are likely to not have permission to publish because the publish action is specific to an individual `wp_navigation` Post which is not created when you insert the Nav block.

This PR proposes another route. We increase the permission gate for the `create` action on the  `wp_navigation` Post type to `edit_theme_options`.

This will effectively mean that _only_ Admins can "create" Navigation posts whilst all other users can only "read". 

This matches nicely with the current `Menus` system in WordPress which is also only accessible to Admin users.

There is [a precedent in Core for adding custom capabilities to "special" post types](https://github.com/WordPress/wordpress-develop/blob/11753b9481676210779bb0622e77c5aa76fabf67/src/wp-includes/post.php#L332). This PR takes the same approach for Navigation posts.

### Why not set _all_ permissions to `edit_theme_options`?

This PR **does not** use the changes from [this Core ticket](https://github.com/WordPress/wordpress-develop/pull/2056) which seeks to alter the perms on the `wp_navigation` post to match those of the Menu post type. Why? Because by adjusting the permissions in this way the [post types endpoint will not show the `wp_navigation` post](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/entities.js#L207-L208) and thus it will not be loaded as an entity in Gutenberg. This will mean [the check for the entity will fail in `getEntityRecord()`](https://github.com/WordPress/gutenberg/blob/1dc4af3d55e637244774fb73f047ca853ecbd748/packages/core-data/src/resolvers.js#L57) before the API request is fired. Therefore we will not be able to "view" the Navigation post.

## ⚠️ String changes post string freeze in WP 5.9 Beta 3

Having spoken with release leads, as requested I have raised [this Trac ticket](https://core.trac.wordpress.org/ticket/54643#ticket). 

## ⚠️ Requires Core patch

As this PR modifies `navigation.php` we will need a corresponding Core patch. The best candidate is probably the one @spacedmonkey already has open in https://github.com/WordPress/wordpress-develop/pull/2056.

Closes https://github.com/WordPress/gutenberg/issues/36286

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

First check you cannot create Navigations as a low permission user:

1. Login as Contributor.
2. Add new Post and add Nav block.
3. See warning that you are not allowed to create Navigation Menus.
4. Check you cannot `Create` a Navigation via buttons in the placeholder.
5. Switch to Admin. Repeat above. See that there is _no_ warning.

Now check you can still view existing Navigations:
1. As a contributor create a Post.
2. Switch to Admin and edit the same Post.
3. Add a Nav block and save.
4. Switch back to Contributor and edit the same post again.
5. You should be able to view the Navigation (you will see a warning about editing not being possible).
6. Check that you _cannot_ see the `Create Menu` option in the `Select Menu` dropdown.

## Screenshots <!-- if applicable -->

<img width="1279" alt="Screen Shot 2021-12-17 at 09 27 10" src="https://user-images.githubusercontent.com/444434/146521702-dd8964b4-5254-4daa-a24a-6c158af9ee95.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Test for publish abtility and display warning

getdave committed:
Update creation of Nav posts to require admin perms

getdave committed:
Only show menu creation option if user has permission to create

getdave committed:
Show warning if unable to create Navigation Menus

getdave committed:
Copy changes from Core patch

See https://github.com/WordPress/wordpress-develop/pull/2056/files

getdave commented on lib/navigation.php:
```diff
@@ -50,6 +50,24 @@ function gutenberg_register_navigation_post_type() {
 			'editor',
 			'revisions',
 		),
+		'capabilities'          => array(
```
Copied from https://github.com/WordPress/wordpress-develop/pull/2056

getdave replied:
I reverted this as it caused too many bugs.

getdave committed:
Only show error if create is not allowed

getdave committed:
Revert "Copy changes from Core patch"

This reverts commit 1872f62a454dc41ce787103e3d1830f7ff00d63c.

getdave commented on lib/navigation.php:
```diff
@@ -50,6 +50,9 @@ function gutenberg_register_navigation_post_type() {
 			'revisions',
 		),
+		'capabilities'          => array(
+			'create_posts' => 'edit_theme_options',
```
Restrict creation of Nav posts to "admin" role level only. Allows us to display a warning immediately upon insertion of Nav block.

Attempting to modify all the permission (as in https://github.com/WordPress/wordpress-develop/pull/2056) will not work because `getEntityRecord` will not return the Navigation Post for viewing.

spacedmonkey commented on lib/navigation.php:
```diff
@@ -50,6 +50,9 @@ function gutenberg_register_navigation_post_type() {
 			'editor',
 			'revisions',
 		),
+		'capabilities'          => array(
+			'create_posts' => 'edit_theme_options',
+		),
```
```suggestion
		'capabilities'          => array(
				// Meta Capabilities.
				'edit_post'              => 'edit_post',
				'read_post'              => 'read_post',
				'delete_post'            => 'delete_post',
				// Primitive Capabilities.
				'edit_posts'             => 'edit_posts',
				'create_posts'           => 'edit_theme_options',
				'edit_others_posts'      => 'edit_theme_options',
				'delete_posts'           => 'edit_theme_options',
				'publish_posts'          => 'edit_theme_options',
				'read_private_posts'     => 'edit_theme_options',
				'read'                   => 'read',
				'delete_private_posts'   => 'edit_theme_options',
				'delete_published_posts' => 'edit_theme_options',
				'delete_others_posts'    => 'edit_theme_options',
				'edit_private_posts'     => 'edit_theme_options',
				'edit_published_posts'   => 'edit_theme_options',
			),
```

@getdave Can you try these values and see if they work for you? I think this line is what was the issue `'edit_posts'             => 'edit_posts',`

spacedmonkey replied:
After more testing, the above does not work.

spacedmonkey replied:
Can you try the following 

```
		'capabilities'          => array(
			'edit_others_posts'      => 'edit_theme_options',
			'delete_posts'           => 'edit_theme_options',
			'publish_posts'          => 'edit_theme_options',
			'create_posts'           => 'edit_theme_options',
			'read_private_posts'     => 'edit_theme_options',
			'delete_private_posts'   => 'edit_theme_options',
			'delete_published_posts' => 'edit_theme_options',
			'delete_others_posts'    => 'edit_theme_options',
			'edit_private_posts'     => 'edit_theme_options',
			'edit_published_posts'   => 'edit_theme_options',
		),
```

getdave replied:
Ok that _seems_ to work so far in my testing.

spacedmonkey commented:
<img width="647" alt="Screenshot 2021-12-17 at 01 01 37" src="https://user-images.githubusercontent.com/237508/146471509-a7ffed36-d2d8-49ee-b198-49bd00a41996.png">

Even if you can't create / edit navigations, you should be able to select one from dropdown.

talldan commented:
This sounds like a good option, and it still leaves the possibility for loosening the restrictions later if the right solution can be found.

Agree with @spacedmonkey that it'd be good to support selecting menus still, but I suppose the dropdown would have to exclude classic menus, as they're converted to a `wp_navigation`.

Code looks good so far.

getdave committed:
Use streamlined permissions

Kudos to @spacedmonkey for https://github.com/WordPress/gutenberg/pull/37454#discussion_r771026149

getdave committed:
Remove inline warning and reenable ability to select existing

getdave committed:
Refactor Notices to reusable hook

getdave committed:
Add notices for creating Menus

getdave committed:
Rename dep for clarity

getdave committed:
Move permission selectors to hook

getdave commented:
> Even if you can't create / edit navigations, you should be able to select one from dropdown.

> Agree with @spacedmonkey that it'd be good to support selecting menus still...

OK I reinstated that feature and created another Notice which appears only when the user is trying to create and there is not existing Nav entity reference.

> ...but I suppose the dropdown would have to exclude classic menus, as they're converted to a wp_navigation.

Is that because Menus are converted into `wp_navigation` at point of selection?

talldan commented:
> OK I reinstated that feature and created another Notice which appears only when the user is trying to create and there is not existing Nav entity reference.

Hopefully we can also hide as many menu creation options as possible.

> Is that because Menus are converted into wp_navigation at point of selection?

Yep exactly.

getdave committed:
Hide other creation options

getdave commented:
I am working on e2e tests now.

getdave committed:
Add e2e test

getdave commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -893,5 +893,28 @@ describe( 'Navigation', () => {
+				`//*[contains(@class, 'components-snackbar__content')][ text()="${ noticeText }" ]`
+			);
+
+			// Expect a console 403 for request to Navigation Areas for lower permisison users.
```
```suggestion
			// Expect a console 403 for request to Navigation Areas for lower permission users.
```

getdave commented:
Just realised I need to hide `Classic` menus from the `Select menu` dropdown.

getdave committed:
Hide classic Menus from dropdown

See https://github.com/WordPress/gutenberg/pull/37454#issuecomment-996569572

getdave committed:
Fix up e2e tests following changes from rebase

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -336,53 +334,30 @@ function Navigation( {
-		}
-
-		if (
+	useNavigationNotice( {
```
My two notes:
* This naming and structure is very specific to navigation, but the implementation is generic
* I had to check how `useNavigationNotice` is implemented to reason about this component. Can we make it obvious?

An implementation like this, could address both and even become an exported effect in 6.0:

```js
const [ showCantEditNotice, hideCantEditNotice ] = useNotice( /*...*/ );
const [ showCantCreateNotice, hideCantCreateNotice ] = useNotice( /*...*/ );
useEffect(() => {
    if( ! isSelected && ! isInnerBlockSelected ) {
        hideCantEditNotice();
        hideCantCreateNotice();
    } else if ( ref ) {
        hideCantCreateNotice();
    }

    if (
       ( isSelected || isInnerBlockSelected ) &&
	hasResolvedCanUserUpdateNavigationEntity &&
    ) {
        if ( ! canUserUpdateNavigationEntity ) {
            showCantEditNotice();
        } else if ( ! canUserCreateNavigation ) {
            showCantCreateNotice();
        }
    }
}, [ /* ...deps ... */]);
```

getdave replied:
Yeh I didn't like the implementation of those hooks. Just rushing really to land the fix. Let me try and refactor...

getdave replied:
Refactored ✅

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -488,6 +463,9 @@ function Navigation( {
 											onClose();
 										} }
 										onCreateNew={ startWithEmptyMenu }
+										canUserCreateNavigation={
```
`showCreate` or `hasCreateNavigation` ? The component doesn't have to "know" the reason for disabling the feature.

getdave replied:
Fixed.

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -642,11 +620,13 @@ function Navigation( {
-						<PlaceholderPreview isLoading />
-					) }
+					{ ! hasResolvedCanUserCreateNavigation ||
+						( ! isEntityAvailable && ! isPlaceholderShown && (
```
This condition is becoming long and hard to unwind. I found it confusing that we're showing a placeholder when `! isPlaceholderShown`. It's not specific to this PR, but it would be nice to have a `<LoadingState />` component or so.

getdave replied:
What should I change here? Or is this for a followup?

adamziel replied:
Nothing to change in this PR, I'm just making a note for the future. I added a mention into the code quality issue https://github.com/WordPress/gutenberg/issues/37190

adamziel commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -31,6 +31,7 @@ const ExistingMenusDropdown = ( {
 	onFinish,
 	menus,
 	onCreateFromMenu,
+	canUserCreateNavigation = false,
```
Same as above, this component doesn't need to know about permissions, I would rename this to `withCreateButton` as it pinpoints the purpose of this argument, whereas `canUserCreateNavigation` could describe many different outcomes.

getdave replied:
Fixed 👍

adamziel commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -102,6 +106,10 @@ export default function NavigationPlaceholder( {
 		blocks,
 		navigationMenuTitle = null
 	) => {
+		if ( ! canUserCreateNavigation ) {
```
I don't like that we have to resort to this kind of defense in depth :( Let's keep it for now – it just makes sense to do it this way in this PR. However, in the future I'd like to get to a place where the flow of data is easy to understand, not error-prone, and we can just be sure this function is not called even once more PRs are merged.

getdave replied:
Agreed.

adamziel commented on packages/block-library/src/navigation/use-navigation-menu.js:
```diff
@@ -79,6 +79,12 @@ export default function useNavigationMenu( ref ) {
 					[ 'delete', 'navigation', ref ]
 				),
+				canUserCreateNavigation:
+					canUser( 'create', 'navigation' ) || undefined,
```
Why `|| undefined`?

getdave replied:
I don't think we need it. Probably some kind of hangover from copy/paste. 

Fixed.

getdave committed:
Update to make component props more agnostic

getdave committed:
Make component props less tighly coupled to permissions

getdave committed:
Remove unneeded undefined fallback

adamziel commented:
It works well, and it is a good start, thank you for working on that @getdave !

There is a bit of a flickering involved:
1. The navigation block is initially in the loading state
2. Then it says the menu is deleted
3. Then it renders, but says one of my blocks is empty
4. Then it renders that block

See the video below:

https://user-images.githubusercontent.com/205419/146559878-4692d9b3-aeff-43d9-88af-4ac6020eed08.mp4

It isn't necessarily a blocker, as this PR solves an important problem, but it would be great to land a fix sooner than later. Maybe [the suspense API support](https://github.com/WordPress/gutenberg/pull/37261#issuecomment-992503841) will help here once it lands? 

There's one other non-blocking rough edge I want to point out: The notice is really easy to miss, and even though it is in place, the sidebar still offers to save the menu. An attempt to save it fails, but 

![CleanShot 2021-12-17 at 15 26 16](https://user-images.githubusercontent.com/205419/146559103-e047424a-ae83-4cb4-be93-1482a4412328.png)
![CleanShot 2021-12-17 at 15 26 23](https://user-images.githubusercontent.com/205419/146559108-ddbe856e-bea7-48b4-94a3-bb522b5709ba.png)

It would be nicer to not display it there in the first place, or make the block immutable (e.g. using a templateLock as @talldan suggested).

adamziel reviewed (approved):
I pointed out some rough edges in this PR, but I don't think any of them is a blocker. The code looks good, the fix works, the problem is important, so I'm approving ✅   I'll defer to @getdave on the best way to apply, postpone, and track the non-blocking issues here.

getdave committed:
Refactor hooks

getdave committed:
Try switch user to admin

getdave commented:
What's left here is trying to figure out why the e2e tests are failing. 

~I suspect it was since the merge of https://github.com/WordPress/gutenberg/pull/37437 as that switched up the before/after hooks which take care of creating and cleaning up the contributor role user.~

~The test that is failing seems to be `allows pages to be created from the navigation block and their links added to menu` which has a potentially unreliable method of handling the 404 on the `url-details` endpoint. I'm now trying mocking the response form that endpoint to try and make that test more resilient.~

Fix is now in https://github.com/WordPress/gutenberg/pull/37501

getdave committed:
Try switching back to admin after each permissions test

getdave committed:
Try targetting test that relies on avoiding a 404 from URL details endpoint

getdave commented:
Please merge https://github.com/WordPress/gutenberg/pull/37501 **before** merging _this_ PR.

getdave commented:
@tellthemachines I believe you will be managing the backporting for Beta 4? If so please can you look to merge this PR as it's important to fix a key permission bug with the Nav block.

Please note that we'll require a companion Trac ticket to alter the `wp_navigation` post type in Core.

🙇 Thank you in advance.
